examples/ExactlyOnce/Program.cs

```diff
@@ -15,788 +15,698 @@
-///     https://github.com/microsoft/FASTER
-///
-///     If you know that the working state will always be less than available memory, you do
-///     not need to use FASTER, or RocksDb or similar - an in memory collection such as
-///     Dictionary<,> will be easier and more performant.
-/// </summary>
+using System.IO;
+
+
 namespace Confluent.Kafka.Examples.ExactlyOnce
 {
     public class Program
     {
-        /// <summary>
-        ///     Default timeout used for transaction related operations.
-        /// </summary>
-        static TimeSpan DefaultTimeout = TimeSpan.FromSeconds(30);
+        private static readonly ILogger logger = Utils.LoggerFactory.CreateLogger<Program>();
 
-        /// <summary>
-        ///     The number of partitions we'll use for each topic.
-        /// </summary>
-        const int NumPartitions = 12;
+        public static async Task Main(string[] args)
+        {
+            var cts = new CancellationTokenSource();
+            var completedCancellationTokenSource = new CancellationTokenSource();
+            Console.CancelKeyPress += (_, e) =>
+            {
+                e.Cancel = true; // prevent the process from terminating.
+                cts.Cancel();
+            };
+            AppDomain.CurrentDomain.ProcessExit += (_, e) =>
+            {
+                cts.Cancel();
+                completedCancellationTokenSource.Token.WaitHandle.WaitOne();
+            };
+
+            if (args.Length != 2)
+            {
+                Console.WriteLine("Usage:");
+                Console.WriteLine("  create_words <transactional-id>");
+                Console.WriteLine("  reverse_words <transactional-id>");
+                completedCancellationTokenSource.Cancel();
+                return;
+            }
+
+            var command = args[0];
+            var transactionalId = args[1];
+
+            IConfiguration configuration = new ConfigurationBuilder()
+                .SetBasePath(Directory.GetCurrentDirectory())
+                .AddJsonFile("./appsettings.json")
+                .Build();
+
+            var adminClientConfig = configuration.GetSection("AdminClient").Get<AdminClientConfig>();
+            using var adminClient = new AdminClientBuilder(adminClientConfig).Build();
+
+            switch (command)
+            {
+                case "create_words":
+                    await new WordCreator().Execute(configuration, transactionalId, adminClient, cts);
+                    break;
+                case "reverse_words":
+                    await new WordReverser().Execute(configuration, transactionalId, adminClient, cts);
+                    break;
+                default:
+                    logger.LogError("Unknown command");
+                    break;
+            }
+
+            completedCancellationTokenSource.Cancel();
+        }
+    }
 
+    public interface ITransactionalProcess<KC, VC, KP, VP>
```

Comment:
To my eye, the current iteration is much better, maybe good enough to go with.

I do think it could be improved though. I wouldn't use interfaces - i would specify the processing method as a delegate, and have config properties/policies to control committing. Making this API flexible gets interesting though - the more flexible you make it the more general / complex the api becomes.

Also, I think the consumer (which may not always be used) should be completely encapsulated in the Processor.